src/socket.cr

```diff
@@ -547,27 +408,9 @@ class Socket < IO
   private def unbuffered_close
     return if @closed
 
-    # Perform libevent cleanup before LibC.close.
-    # Using a file descriptor after it has been closed is never defined and can
-    # always lead to undefined results. This is not specific to libevent.
     @closed = true
-    evented_close
-
-    # Clear the @volatile_fd before actually closing it in order to
-    # reduce the chance of reading an outdated fd value
-    _fd = @volatile_fd.swap(-1)
```

Comment:
Hmm. After looking at evented_close, I'm wondering a bit if it would make sense to split that method in two, one part that removes all pending events that would happen before closing, and one that wake up the result, called after everything. As `@closed` isn't volatile it seems strange to wake up stuff (that can be run in other threads) before we actually is done. Setting `@closed` to false like that is definitely safe in the single threaded mode, but in multithreaded I wonder how far the assignment to `@closed` may be moved around by the compiler and CPU.

Reply:
Sounds legit, but I'd prefer to leave such restructuring for a follow-up, to keep the complexity of this simple yet extensive refactoring low.